src/BSON/Iterator.c

```diff
@@ -263,10 +298,92 @@ static HashTable* php_phongo_iterator_get_properties(phongo_compat_object_handle
 	return php_phongo_iterator_get_properties_hash(object, false);
 }
 
+/* Iterator handlers */
+static void php_phongo_iterator_it_dtor(zend_object_iterator* iter)
+{
+	zval_ptr_dtor(&iter->data);
+}
+
+static int php_phongo_iterator_it_valid(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	return intern->valid ? SUCCESS : FAILURE;
+}
+
+static zval* php_phongo_iterator_it_get_current_data(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	return php_phongo_iterator_get_current(intern);
+}
+
+static void php_phongo_iterator_it_get_current_key(zend_object_iterator* iter, zval* key)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	if (!php_phongo_iterator_key(intern, key)) {
+		// Exception already thrown
+		return;
+	}
+}
+
+static void php_phongo_iterator_it_move_forward(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	php_phongo_iterator_next(intern);
+}
+
+static void php_phongo_iterator_it_rewind(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	php_phongo_iterator_rewind(intern);
+}
+
+static HashTable* php_phongo_iterator_it_get_gc(zend_object_iterator* iter, zval** table, int* n)
+{
+	*n     = 1;
+	*table = &iter->data;
+	return NULL;
+}
+
+static const zend_object_iterator_funcs php_phongo_iterator_it_funcs = {
+	php_phongo_iterator_it_dtor,
+	php_phongo_iterator_it_valid,
+	php_phongo_iterator_it_get_current_data,
+	php_phongo_iterator_it_get_current_key,
+	php_phongo_iterator_it_move_forward,
+	php_phongo_iterator_it_rewind,
+	NULL,
```

Comment:
This is the function pointer for `invalidate_current`, which isn't used. Quoting from [phpinternalsbook](https://www.phpinternalsbook.com/php5/classes_objects/iterators.html):
> The only handler that has no correspondence in userland is `invalidate_current`, which can be used to destroy the current key/value. The handler is largely unused though, in particular `foreach` won’t even call it.

We could point this to `php_phongo_iterator_free_current`, although I'm not sure what the benefits of that would be.

Reply:
`php_phongo_iterator_free_current` is already called from the necessary places.

I'm fine with leaving `invalidate_current` unset, unless there's prior art for other iterator implementations using it.

Reply:
Added the comment as suggested.